### PR TITLE
feat/node: send Ack on consensused message

### LINKED ISSUE #1666: On consensus on a message, Send AckMessage
Dependent on https://github.com/maidsafe/routing/issues/1668 and RMD impl: https://github.com/maidsafe/routing/projects/1

### PR DESCRIPTION
**Pull request by maqi**
Closes #1666

### CODE REVIEW

File: src/states/elder/mod.rs [hunk 1]
```diff
@@ -2083,14 +2084,25 @@ impl Approved for Elder {
 
         self.update_peer_states(outbox);
 
-        if self_sec_update {
+        let dst = if self_sec_update {
             self.chain.reset_candidate_if_member_of(sec_info.members());
             self.send_neighbour_infos();
+
+            Authority::Section(*self.name())
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: Why do you send that Ack message if this section is ourselves?

File: src/states/elder/mod.rs [hunk 2]
```diff
@@ -2083,14 +2084,25 @@ impl Approved for Elder {
...
             self.remove_from_proving_section_cache(&sec_info);
             // Vote for neighbour update if we haven't done so already.
             // vote_for_event is expected to only generate a new vote if required.
-            self.vote_for_event(sec_info.into_network_event());
+            self.vote_for_event(sec_info.clone().into_network_event());
+
+            Authority::Section(sec_info.prefix().name())
+        };
+
+        if !is_initial {
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: Do we really need this `is_initial`?
When it is true it seem to be for our own section?

File: src/states/elder/mod.rs [hunk 3]
```diff
@@ -677,6 +681,12 @@ impl Elder {
         }
     }
 
+    fn handle_section_info_ack(&mut self, sec_info: &SectionInfo, _src: XorName, dst: XorName) {
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: This cannot be called when receiving a message as we are updating shared_state. (can only be done on consensus).
BTW, this is part of the other issue to handle it.

File: src/states/elder/mod.rs [hunk 4]
```diff
@@ -2090,7 +2100,16 @@ impl Approved for Elder {
             self.remove_from_proving_section_cache(&sec_info);
             // Vote for neighbour update if we haven't done so already.
             // vote_for_event is expected to only generate a new vote if required.
-            self.vote_for_event(sec_info.into_network_event());
+            self.vote_for_event(sec_info.clone().into_network_event());
+        }
+
+        if !self_sec_update {
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: Why make it a separate if condition.
If we want clarity, probably better to create a sub-function, maybe `self.send_section_info_ack(sec_info);`

File: src/states/elder/mod.rs [hunk 5]
```diff
@@ -2090,7 +2100,16 @@ impl Approved for Elder {
...
+            // Sending out Ack message.
+            let src = Authority::Section(*self.name());
+            let dst = Authority::Section(sec_info.prefix().name());
+            let content = MessageContent::SectionInfoAck(sec_info);
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: I would probably stick to `AckMessage` as in the issue unless there is a good reason.

File: src/states/elder/mod.rs (hunk 3 above)
Comment (jeanphilippeD) [reply src/states/elder/mod.rs]: What this mean is that function should not be in that PR, so is update_their_keys.

Comment (maqi) [reply src/states/elder/mod.rs]: Then in this PR, the `receive` can be totally skipped?

Comment (jeanphilippeD) [reply src/states/elder/mod.rs]: Yes, i.e you don't have to do anything useful with the receive.
Looking at your test code, maybe you could replace your `sent_section_info_ack` with `received_section_info_ack`. and setting it to true is the only thing this PR does.

Comment (maqi) [reply src/states/elder/mod.rs]: hmm...   here we only `sent out section info ack`, and hasn't receive that yet.
so, I think it shall not be `received_section_info_ack` ?
or, you mean `received_section_info` ?

Comment (jeanphilippeD) [reply src/states/elder/mod.rs]: I meant something like replace
```
            (AckMessage(sec_info), Section(src), Section(dst)) => {
                self.handle_section_info_ack(&sec_info, src, dst);
                Ok(())
            }
```

with 
```
            (AckMessage(sec_info), Section(src), Section(dst)) => {
                self.received_section_info_ack = true;
                Ok(())
            }
```

Comment (maqi) [reply src/states/elder/mod.rs]: Ah, ok

File: src/chain/chain.rs [hunk 6]
```diff
@@ -526,6 +526,11 @@ impl Chain {
         Ok(())
     }
 
+    /// Update other sections' public keys.
+    pub fn update_their_keys(&mut self, sec_info: &SectionInfo) {
```
Comment (jeanphilippeD) [inline src/chain/chain.rs]: Not used, should be in follow up PR?

File: src/chain/shared_state.rs [hunk 7]
```diff
@@ -156,6 +156,13 @@ impl SharedState {
                 .collect()
         }
     }
+
+    /// Update other sections' public keys.
+    pub fn update_their_keys(&mut self, sec_info: &SectionInfo) {
```
Comment (jeanphilippeD) [inline src/chain/shared_state.rs]: Not used, should be in follow up PR?

**Comment by jeanphilippeD** [comment]
Looks good, a minor comment

File: src/states/common/approved.rs [hunk 8]
```diff
@@ -59,6 +59,7 @@ pub trait Approved: Relocated {
     fn handle_neighbour_merge_event(&mut self) -> Result<(), RoutingError>;
 
     /// Handles an accumulated `SectionInfo` event.
+    /// If triggered during initial, an ack message shall not be sent out.
```
Comment (fizyk20) [inline src/states/common/approved.rs]: Could you rephrase the comment? I'm not sure what "triggered during initial" means.
